File: server/src/middleware/error.middleware.js
import { Logger } from '/app/common/logger.js';
import { Definitions } from '/app/common/definitions.js';

const definitions = new Definitions();

const logger = new Logger({
    serviceName: 'error-middleware',
    logLevel: definitions.server.logLevel || 'INFO'
});

export const errorHandler = (err, req, res, next) => {
    // Log the full error stack in development
    if (definitions.server.node_env === 'development') {
        logger.error(`Full error stack: ${err.stack}`);
    }

    // Log the error message
    logger.error(`Error occurred: ${err.message}`);

    if (err.name === 'ValidationError') {
        logger.warn(`Validation Error: ${err.message}`);
        return res.status(400).json({
            error: 'Validation Error',
            details: err.message
        });
    }

    if (err.name === 'UnauthorizedError') {
        logger.warn(`Unauthorized Access: ${err.message}`);
        return res.status(401).json({
            error: 'Unauthorized',
            details: err.message
        });
    }

    // Generic server error
    res.status(500).json({
        error: 'Internal Server Error',
        details: definitions.server.node_env === 'development' ? err.message : 'An unexpected error occurred'
    });
};

File: server/src/middleware/error_middleware.py
import os
from flask import jsonify
from common.logging import logger


def error_handler(error):
    """Global error handler for the application"""

    # Get error details
    status_code = getattr(error, 'code', 500)
    message = str(error)

    # Log the error
    logger.error(f'Error: {message}')
    if status_code == 500:
        logger.error(f'Stack trace: {error.__traceback__}')

    # Return error response
    response = {
        'error': message,
        'status_code': status_code
    }

    # Add stack trace in development
    if os.getenv('FLASK_ENV') == 'development':
        response['stack_trace'] = str(error.__traceback__)

    return jsonify(response), status_code
